Feature prompt on launch on templates

john-westcott-iv opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Enable prompting on launch of the following items in Job Templates:
* Execution Environment
* Labels
* Forks
* Job Slice Count
* Timeout
* Instance Groups

Enable prompting on launch for the following items in Workflow Job Templates:
* Labels
* Skip Tags
* Job Tags


<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - UI
 - Collection
 - CLI

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 21.5.1.dev132+g2dfdf504f8
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

Commit:
adding prompt-to-launch field on Labels field in Workflow Templates; with necessary UI and testing changes

Co-authored-by: Keith Grant <keithjgrant@gmail.com>

mabashian committed:
Refactors EE Lookup to support prompting.  Adds prompting for EE to JT form

Adds prompt on launch buttons to labels, forks, job slicing, timeout, and instance groups

Adds prompting for labels on workflow job template

Updates flags that denote when prompting is necessary in various places

Adds prompting support for timeout, job slicing, forks, labels, instance groups and execution environments to the prompt details

Show prompted ee, forks, job slice and labels on schedule details

Adds support for ee, labels, forks, job slicing and timeout prompting to the node view modal

Add default values when prompting for ee's, forks, job slicing and timeout

Adds launch prompt step for execution environments

Adds fields for timeout, job slicing and forks to other prompts step of launch

john-westcott-iv committed:
JT param everything (#12646)

* Making almost all fields promptable on job templates and config models
* Adding EE, IG and label access checks
* Changing jobs preferred instance group function to handle the new IG cache field
* Adding new ask fields to job template modules
* Address unit/functional tests
* Adding migration file

mabashian committed:
Change ask_job_slicing_on_launch to ask_job_slice_count_on_launch to match api

Adds support for prompting labels on launch in the UI

Fix execution environment prompting in UI

Round out support for prompting all the things on JT launch

Adds timeout to job details

Adds fetchAllLabels to JT/WFJT data models

Moves labels methods out to a mixin so they can be shared across JTs/WFJTs/Schedules

Fixes bug where ee was not being sent on launch

Adds the ability to prompt for ee's, ig's, labels, timeout and job slicing to schedules

Fixes bug where saving schedule form without opening the prompt would throw errors

Adds support for IGs and labels to workflow node prompting

Adds support for label prompting to node modal

Fix job template form tests

john-westcott-iv committed:
Adding EE/IG/labels/forks/timeout/job_slice_count to schedules

Modifying schedules to work with related fields

Updating awx.awx.workflow_job_template_node

mabashian committed:
Leverage the IG mixin on the schedules model

Move associate/disassociate label methods into mixin

Move label/IG saving out to related endpoints off of a schedule

mabashian committed:
Cleanup UI linting, tests, and import

Cleans up UI linting errors

Fix broken UI unit tests

Adds missing LabelsMixin import

obaranov committed:
Add awxkit optional fields for new prompts

Added additional fields for the awskit to support prompts:
 * ee
 * labels
 * forks
 * timeout
 * ig
 * job_slices

john-westcott-iv committed:
Adding unit/functional tests, fixing tests

Making common class for LabelList

Fixing related field name

Fixing get_effective_slice_ct to look for corerct field and also override _eager_field

mabashian committed:
Fixes for various prompt related ui issues

Fixes bug where Forks showed up in both default values and prompted values in launch summary

Fixes prompting IGs with defaults on launch

Make job tags and skip tags full width on workflow form

Fixes bug where we attempted to fetch instance groups for workflows

Fetch default instance groups from jt/schedule for schedule form prompt

Grab default IGs when adding a node that prompts for them

Adds support for saving labels on a new wf node

Fix linting errors

Fixes for various prompt on launch related issues

Adds support for saving instance groups on a new node

Adds support for saving instance groups when editing an existing node

Fix workflowReducer test

Updates useSelected to handle a non-empty starting state

Fixes visualizerNode tests

Fix visualizer test

Second batch of prompt related ui issues:

Fixes bug saving existing node when instance groups is not promptable

Fixes bug removing newly added label

Adds onError function to label prompt

Fixes tooltips on the other prompts step

Properly fetch all labels to show on schedule details

mabashian committed:
Extends LabelSelect to have a custom chip render.  This allows us to disable labels that cannot be removed on job launch

AlanCoding committed:
Add WorkflowJob.instance_groups and distinguish from char_prompts

This removes a loop that ran on import
  the loop was giving the wrong behavior
  and it initialized too many fields as char_prompts fields

With this, we will now enumerate the char_prompts type fields manually

john-westcott-iv committed:
Making labels additive and not adding a many item to config if already in parent

AlanCoding committed:
DRY edits to access classes for new prompts

Remove if-not-data conditional from WFJTnode.can_change
  these are cannonical for can_add, but this looks like a bug

Change JTaccess.can_unattach to call same method in super()
  previously called can_attach, which is problematic

Better consolidate launch config m2m related checks

Test and fix pre-existing WFJT node RBAC bug

recognize not-provided instance group list on launch, avoiding bug where it fell back to default

fix bug where timeout field was saved on WFJT nodes after creating approval node

remove labels from schedule serializer summary_fields

remove unnecessary prefetch of credentials from WFJT node queryset

john-westcott-iv committed:
Changing migration sfrom 0167 to 0168

Fixing linting error

mabashian committed:
Adds elements and identifiers for cypress tests

Properly display instance groups and labels on node details view

AlexSCorey commented:
Need QE re: @mabashian

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3035,6 +3041,9 @@ class Meta:
             'webhook_guid',
+            # TODO: Do we want these here or just in the summary fields?
+            'labels',
+            'instance_groups',
```
I think delete. I'm not even really sure of the effect right now.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3061,6 +3070,11 @@ def get_related(self, obj):
+            if obj.instance_groups:
+                res['instance_groups'] = self.reverse('api:job_instance_group_list', kwargs={'pk': obj.pk})
+        except ObjectDoesNotExist:
+            pass
```
I don't think this is the right way to use `ObjectDoesNotExist`, because it's a relationship, and `obj.instance_groups` should always exist, so better to get rid of the try-except.

john-westcott-iv replied:
I think initially I was thinking incase an IG got deleted but since they are FK constraints that wouldn't matter, updated.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3443,12 +3467,13 @@ def get_related(self, obj):
     def validate_extra_vars(self, value):
         return vars_validate_or_raise(value)
 
+    # posting
```
delete comment

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3609,6 +3639,10 @@ class LaunchConfigurationBaseSerializer(BaseSerializer):
     skip_tags = serializers.CharField(allow_blank=True, allow_null=True, required=False, default=None)
     diff_mode = serializers.BooleanField(required=False, allow_null=True, default=None)
     verbosity = serializers.ChoiceField(allow_null=True, required=False, default=None, choices=VERBOSITY_CHOICES)
+    execution_environment = serializers.PrimaryKeyRelatedField(queryset=ExecutionEnvironment.objects.all(), required=False, allow_null=True, default=None)
```
This follows the same pattern as `inventory`, and I don't think it needs to be defined on the serializer, because the DRF serializer does everything we need for this.

john-westcott-iv replied:
Removed.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3624,13 +3658,21 @@ class Meta:
+        res['labels'] = self.reverse('api:{}_labels_list'.format(get_type_for_model(self.Meta.model)), kwargs={'pk': obj.pk})
+        res['instance_groups'] = self.reverse('api:{}_instance_groups_list'.format(get_type_for_model(self.Meta.model)), kwargs={'pk': obj.pk})
+        if obj.execution_environment_id:
+            res['execution_environment'] = self.reverse('api:execution_environment_detail', kwargs={'pk': obj.execution_environment_id})
```
the ordering is probably inconsistent. It would be good if we could formalize the order of the 5 - inventory, EE, labels, credentials, IGs

john-westcott-iv replied:
Order updated.

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -3640,7 +3682,11 @@ def _build_mock_obj(self, attrs):
         field_names = set(field.name for field in self.Meta.model._meta.fields)
         for field_name, value in list(attrs.items()):
-            setattr(mock_obj, field_name, value)
+            if field_name == 'execution_environment':
```
I still can't figure out why the EE should have an exception case here.

john-westcott-iv replied:
Removed.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -208,6 +209,26 @@ def api_exception_handler(exc, context):
+        # If a label already exists in the database, attach it instead of erroring out
+        # that it already exists
+        if not getattr(self, 'label_filter', None):
+            return Response(dict(msg=_('Class {} missing label filter.'.format(self.__class__.__name__))), status=status.HTTP_400_BAD_REQUEST)
```
Again, I would like to touch this up as this is never an expected response.

mabashian committed:
Add more ui unit test coverage for prompt changes

Flips default job/skip tags value from empty string to null on WF form

AlanCoding commented on awx/main/models/jobs.py:
```diff
@@ -965,8 +1017,12 @@ class Meta:
     credentials = models.ManyToManyField('Credential', related_name='%(class)ss')
+    labels = models.ManyToManyField('Label', related_name='%(class)s_labels')
+    execution_environment = models.ForeignKey(
+        'ExecutionEnvironment', null=True, blank=True, default=None, on_delete=polymorphic.SET_NULL, related_name='%(class)s_as_prompt'
```
As this is a `ForeignKey`, not a many-related field, it would be really nice to have help_text for OPTIONS, especially to echo the adherence to its `ask_` rule.

john-westcott-iv replied:
This field is on the JobLaunchConfig the JobTemplate endpoint gets the POST options from ExecutionEnvironmentMixin. Is there somewhere in the API that you can see job launch configs?

AlanCoding replied:
Schedules and WFJT nodes _are_ `JobLaunchConfig`s. It is an abstract model, but that just means that it has content shared by other concrete models. Those I list are not the _only_ concrete model that subclasses this and inherits its fields, but they are the easiest examples because they are listed and editable in the API the same way other normal resources are.

john-westcott-iv replied:
Good call on the inheritance from schedules. The schedules endpoint was not display help text. That has been resolved.

AlanCoding commented on awx/main/tests/functional/models/test_job_launch_config.py:
```diff
@@ -61,7 +62,42 @@ def test_char_field_only(self, config_factory):
+        }
         config = config_factory(prompts)
+        print(prompts)
+        print(config.prompts_dict())
```
delete print statements

AlexSCorey commented on awx/ui/src/components/LaunchPrompt/steps/OtherPromptsStep.js:
```diff
@@ -22,72 +24,105 @@ const FieldHeader = styled.div`
 `;
 
 function OtherPromptsStep({ launchConfig, variablesMode, onVarModeChange }) {
+  const helpTextSource = launchConfig.job_template_data
```
This is neat

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -208,6 +209,26 @@ def api_exception_handler(exc, context):
     return exception_handler(exc, context)
 
 
+class LabelList(DeleteLastUnattachLabelMixin, SubListCreateAttachDetachAPIView):
```
@john-westcott-iv this has a re-definition of `LabelList` at line 4501.

that should have been picked up by the linter, but we selectively enable rules b/c our first line of defense is black. But re-definition is a big one that we should capture.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -617,6 +638,23 @@ class ScheduleCredentialsList(LaunchConfigCredentialsBase):
     parent_model = models.Schedule
 
 
+class ScheduleLabelsList(LabelList):
```
This view is a sublist-create-attach view. That means that you can create a label by

POST to `/api/v2/schedule/N/labels/`

@mabashian does the UI use this to create labels? Or does it go create them by POST to `/api/v2/labels/`? If the UI does not use it, we should change the views from a create-attach-sublist to just an attach-sublist type view. Otherwise we have this un-tested creation mechanism hanging around, and a multitude of things may be wrong about it.

~Same goes for the instance groups sublist.~

EDIT: instance groups sublist does not allow creating an instance group, so that's what we want. :+1: to that.

EDIT: the answer I am presenting in the related PR is that label sublists _should_ allow creating labels, and I am consolidating the view code for this so WFJT node & schedule label-lists work the same as other editable label-lists.

AlanCoding commented on awxkit/awxkit/api/pages/workflow_job_templates.py:
```diff
@@ -42,6 +42,9 @@ def payload(self, **kwargs):
+            "ask_tags_on_launch",
             "limit",
             "scm_branch",
             "survey_enabled",
```
Action item:

Also in this module, the WFJT model has a `add_label` method. The WFJT node should _also_ have this method, because this PR adds labels on it.

Same goes for labels on schedules, the awxkit Schedule model should get a new `add_label` method.

----

Instance groups was also an added endpoint, and `add_instance_group` is a method on some models. This method should be added to WFJT nodes and schedules as well.

obaranov replied:
I'll do that

obaranov replied:
add_label PR: https://github.com/ansible/awx/pull/12886

john-westcott-iv replied:
Resolving from merge

obaranov committed:
Add new add_label method to the wfjt node and schedules awxkit models

obaranov committed:
Added instance_groups method to the awxkit models

Also added additional payload fields to the wfjt model.

john-westcott-iv committed:
Removing non-functional lines

john-westcott-iv committed:
Removing labels and instance_groups from the job serializer page as top level items (still in summary fields)

sarabrajsingh committed:
adding fix for labels pushdown on workflow job nodes

john-westcott-iv committed:
Fixing serializers per review

Removing try/except around instance_groups

Removing redefined execution_environment

Reordering labels/creds/igs/ee/etc

Removing special treatment for EEs when doing setattrs

Adding help_text to execution environments

Adding EE serializer on JobCreateScheduleSerializer

obaranov commented:
I have test failures with /api/v2/jobs/142/create_schedule endpoint (test_prompt_ee_on_schedule test). I start job with an execution environment prompt. And then calling create_schedule produces the

`tools_awx_1     | TypeError: Object of type ExecutionEnvironment is not JSON serializable`

Here is simple pseudo code to reproduce: 
```

        jt = factories.job_template(ask_execution_environment_on_launch=True)
        job = jt.launch(dict(execution_environment=ee_copy.id)).wait_until_completed()
        create_schedule = job.related.create_schedule.get()
        assert create_schedule.prompts == dict(execution_environment=ee_copy.id)
        schedule = create_schedule.post()
```
The same happens for instance group prompt.

AlanCoding committed:
Complete consolidation of the label views

john-westcott-iv commented:
We need to review `is_detached`, `get_ophaned_labels` and `is_candidate_for_detach` from the Labels model to make sure they account for the new relations to schedules and wfjtnodes and any other objects.

sarabrajsingh commented:
https://github.com/ansible/awx/pull/12888 - needs to be scoped in here as well

john-westcott-iv committed:
Changing label functions to account for new relationships

Removing unreferenced get_orphaned_labels

Forcing forks and job_slice_count to be >=0

AlanCoding committed:
Add support for instance_groups and labels on schedule create

keithjgrant commented on awx/ui/src/components/LaunchButton/LaunchButton.js:
```diff
@@ -60,6 +71,27 @@ function LaunchButton({ resource, children }) {
+        } = await JobTemplatesAPI.readInstanceGroups(resource.id);
+
+        setInstanceGroups(results);
+      }
```
These three requests (`readSurvey`, `readLabels` and `readInstanceGroups`) are made in serial. It would be more efficient to get both promises before awaiting the results (though it will require extra `if` blocks at the end before setting state vars).

mabashian replied:
👍

keithjgrant commented on awx/ui/src/components/LaunchPrompt/LaunchPrompt.js:
```diff
@@ -53,6 +62,61 @@ function PromptModalForm({
+      const labelIds = [];
+      const newLabels = [];
+      const labelRequests = [];
+      let organizationId = resource.organization;
+      values.labels.forEach((label) => {
+        if (typeof label.id !== 'number') {
+          newLabels.push(label);
+        } else {
+          labelIds.push(label.id);
+        }
+      });
+
+      if (newLabels.length > 0) {
+        if (!organizationId) {
+          // eslint-disable-next-line no-useless-catch
+          try {
+            const {
+              data: { results },
+            } = await OrganizationsAPI.read();
+            organizationId = results[0].id;
+          } catch (err) {
+            throw err;
+          }
+        }
+      }
+
+      newLabels.forEach((label) => {
+        labelRequests.push(
+          LabelsAPI.create({
+            name: label.name,
+            organization: organizationId,
+          }).then(({ data }) => {
+            labelIds.push(data.id);
+          })
+        );
+      });
+
+      await Promise.all(labelRequests);
+
+      setValue('labels', labelIds);
```
Consider moving most or all of this into a helper function to keep `handleSubmit` down to a scanable length

mabashian replied:
Good call - there's identical code in the Schedule Edit component so I'll pull this all out and put it somewhere where it can be imported.

AlanCoding committed:
Write logic to combing workflow labels, IGs with nodes

Additionally, move the inventory-specific hacks of yesteryear
  into the prompts_dict method of the WorkflowJob model
  try to make it clear exactly what this is hacking and why

Correctly summarize label prompts, and add missing EE

Expand unit tests to apply more fields

adding missing fields to preserve during copy to workflow.py

Fix bug where empty workflow job vars blanked node vars (#12904)

* Fix bug where empty workflow job vars blanked node vars

* Fix bug where workflow job has no extra_vars, add test

* Add empty workflow job extra vars to assure fix

AlanCoding commented on awx/main/models/jobs.py:
```diff
@@ -314,10 +331,16 @@ def get_effective_slice_ct(self, kwargs):
+            actual_slice_count = kwargs['job_slice_count']
+            # Set the eager fields if its there as well
+            if '_eager_fields' in kwargs and 'job_slice_count' in kwargs['_eager_fields']:
+                kwargs['_eager_fields']['job_slice_count'] = actual_slice_count
```
As a coding practice... obviously I would not like a `get_*` method to _set_ something like this.

But it's much more troublesome that there was some reason that caused this to need to get set here in the first place. This would seem to indicate that `job_slice_count` doesn't get copied on launch to the job. This should happen in `create_unified_job` irrelevant of the prompts.

```
In [8]: set(JobTemplate.get_ask_mapping().keys()) - set(WorkflowJobTemplate._get_unified_job_field_names())
Out[8]: 
{'credentials',
 'diff_mode',
 'execution_environment',
 'forks',
 'instance_groups',
 'job_slice_count',
 'job_type',
 'timeout',
 'verbosity'}
```

I expect some of these probably need to be added to the workflow `_get_unified_job_field_names`, because they are valid workflow fields.

john-westcott-iv committed:
Altering --timeout from awxkit to --action-timeout to remove conflict with new launch timeout

john-westcott-iv committed:
Get more specific as to which timeout caused the issue

AlanCoding commented on awx/main/utils/common.py:
```diff
@@ -531,7 +531,16 @@ def copy_m2m_relationships(obj1, obj2, fields, kwargs=None):
+                        # Labels are additive so we are going to add any src labels in addition to the override labels
+                        if field_name == 'labels':
+                            for jt_label in src_field_value.all():
+                                getattr(obj2, field_name).add(jt_label.id)
```
Merging of credentials in this process is handled here:

https://github.com/ansible/awx/blob/063b0d08343938295a3c8f6ccb5342ffd2421d40/awx/main/models/unified_jobs.py#L402-L409

Yes, all 3 relationships, instance_groups, labels, and credentials have distinct rules associated with them. But handing 1 relationship _before_ the method `copy_m2m_relationships` call, and 2 relationships _inside_ the method is going to make this much harder for anyone to ever follow in the future.

To pick one, I would say it should be done before the method call, because that can be adjusted to be more testable. That still leaves you with an exception case for _saving_ instance groups, which would have to remain here.

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -973,22 +977,42 @@ def create_config_from_prompts(self, kwargs, parent=None):
+                    for item in job_item:
+                        # Do not include this item in the config if its in the parent
+                        if item not in parent_items:
+                            getattr(config, field_name).add(item)
```
The "Ordered field" block here can't be right - it contradicts the prompting rules we wrote for `instance_groups`.

If the launch-time list is not `[]`, then _the entire list_ constitutes the prompted value. This loop obtains any items _added relative to the JT_ and that's completely irrelevant of what it actually does. _Even if the entire list matches the JT_, if you provide it on launch, that gives a different meaning. The JT `instance_groups` says "use these groups, but fallback to inventory/org". The prompted list says "_only_ use these groups".

Putting it that way, it's also clear that the `default` value offered in the `JobLaunchSerializer` is wrong, because it returns the JT instance groups. Using that as-is would use that same list, but with a different meaning to it. So either the entire construct of the `default` should be declared invalid for this field, or the view should present `"default": []` or `"default": null`, because this is what obtains parity with the basic JT behavior.


While the prompts store is only used for the obscure purposes of relaunch or the "create_schedule", this makes obvious another more likely and severe bug. The method `create_unified_job` takes `kwargs` with an unclear meaning. Even if the UI gives `[]`, that doesn't mean we're off the hook because `copy_m2m_relationships` takes _any_ value in kwargs to be an _override_, including a falsy value, so it's not even clear that our our intended not-provided value will leave the JT values untouched.

I'm sure testing is at high risk of missing these things, because we are very light on testing (if at all) filling in JT values and then assuring launch works _without_ providing prompts, when prompts are accepted. Even if we did, integration tests use its own client, and it's not clear even what payload to the launch POST they are even _supposed_ to provide for this case, because in the request data, you can _exclude_ the key, set to _null_, set to `[]`, or set to the "default" values. I don't think we have ever touched that testing matrix.

AlanCoding replied:
I don't want to leave this too vague...

I'm saying that, as you have already added special logic in `copy_m2m_relationships`, you need more special logic. Before you start the loop `for ig_id in override_field_val:` in the special IG logic, it needs to be checked that the data constitute an override in the first place.... like, indent that into a `if override_field_val:` conditional. But I'd hate to jump right into that without first characterizing relevant test cases.

AlanCoding replied:
Oh, and this is actually impossible / impractical to test in integration tests. Only unit tests, because the resolved list is never reported by the API.

AlanCoding replied:
Scratch some of that, don't take my prior advice. In the event that IGs are not provided, IIRC, we do not want to copy the JT values at all. Again, this is philosophically very problematic with `copy_m2m_relationships`, and it should probably be done before that, but using the existing code structure, that method should _always_ `continue` on the instance_groups field to avoid a copy of the JT list. I do need to articulate a specific test case for this.

AlanCoding replied:
We discussed this and I am developing a potential patch.

https://github.com/ansible/awx/pull/12925

Needs work post-migration merge.

obaranov committed:
Add additional schedule fields for new prompts

sarabrajsingh committed:
adding test coverage to ensure that FIELDS_TO_PRESERVE_AT_COPY is behaving as expected for WFJTs

AlanCoding committed:
Fix another bug applying extra_vars to incompatible job types

john-westcott-iv committed:
Fixing is_detached methods' filters

AlanCoding commented on awx/main/tests/functional/api/test_job_runtime_params.py:
```diff
@@ -157,11 +189,28 @@ def test_job_accept_empty_tags(job_template_prompts, post, admin_user, mocker):
             post(reverse('api:job_template_launch', kwargs={'pk': job_template.pk}), {'job_tags': '', 'skip_tags': ''}, admin_user, expect=201)
             assert JobTemplate.create_unified_job.called
-            assert JobTemplate.create_unified_job.call_args == ({'job_tags': '', 'skip_tags': ''},)
+            assert JobTemplate.create_unified_job.call_args == ({'job_tags': '', 'skip_tags': '', 'forks': 1, 'job_slice_count': 0},)
```
@john-westcott-iv going through tests, I was surprised by these numbers. This doesn't look desirable, I would prefer these files to default to `None` / not-provided.

AlanCoding replied:
addressed in https://github.com/ansible/awx/pull/12917

mabashian committed:
Remove duplicate Limit detail on schedule

Bumps migration number from 0168 to 0169

Make labels and IGs requests synchronously when getting launch data

Moves label creation out to a util

AlanCoding committed:
Treat instance_groups prompt as template-less

mabashian committed:
Removes fetching of default instance groups in the UI on launch and schedule/node creation

AlanCoding committed:
Fix bug with missing parent field and diff with parent

Remove corresponding views for job instance_groups

Validate job_slice_count in API

Remove defaults from some job launch view prompts
  the null default is preferable

obaranov committed:
Ensure schedule collection test has enough hosts for slices

john-westcott-iv committed:
Collection test fixes from prompting changes

DNE can sometimes be dne depending on versions, fixing test to find either

Adding additional node to Demo Inventory for job slice counting

AlanCoding commented:
https://github.com/ansible/awx/compare/devel...AlanCoding:squash_attempt3?expand=1

 if you want squashed
